### ISSUE TITLE
Improving the voting gui

### ISSUE THREAD

**Issue by hiker** [supertuxkart/stk-code#3649]
As discussed on irc today, we agreed to try to implement Deve's track-selection+voting gui:
![voting gui](https://user-images.githubusercontent.com/6138652/50001849-46375500-fff2-11e8-9684-4f93b574519e.png)

Adding a progress bar to the kart and track screens (one timeout period for both).

I'll start with updating the voting_gui branch, since this has already most of the infrastructure implemented (handling votes by using track/lap/reverse from one vote only), floating-point based progress screen, updating an animation while loading), which should be finished in a day or two.

**Comment by hiker** [comment]
@benau, @deveee, @auriamg fyi.

**Comment by Benau** [comment]
can u make the kart selection screen start showing progress bar only 1st player vote so we don't need to raise network protocol

**Comment by hiker** [comment]
@benau, I already told you on irc that the players we asked on irc agreed that it is easier to understand if the timing starts from the beginning. I don't understand what the problem is to increase the protocol number? I expect this to happen a couple of time still before the RC even (e.g. we might start to add compression and maybe even increase state frequency, not to mention fixing bugs). Am I missing something?

**Comment by Alayan-stk-2** [comment]
Having the timer start from the beginning is useful for consistency. This won't make the process too long if people take their time to submit a first choice, but with a decent value (I think 30s for both kart selection + track selection would be good, 20s would usually be too short) the player will have enough time without being hardpressed to do a choice quickly.

Also otherwise, you'd have the progress bar full, and then suddenly it starts decreasing while you're still selecting your kart ? This wouldn't be a good UI.

Updating the network protocol before the beta shouldn't be a big issue imo.

About the UI discussed at the beginning : as said on the IRC, slightly smaller track thumbnails allowing to display a 4x4 grid instead of a 3x3 grid should help a lot to quickly find a track.

**Comment by Benau** [comment]
ok

**Comment by Benau** [comment]
![test](https://user-images.githubusercontent.com/3252341/50013256-9b567500-ffb8-11e8-876d-dc9628aa47d3.jpg)
(the icon can be improved later)
alayan said it "messy" to have icon with text mixed with header so what about we use icon for all?
and i can make an icon for random item location, and use soccer ball for goal limit and clock (loading.png) icon for time limit, then you should have enough space for player name

and i think we only need highlight after fixing the update-gui-loading branch, and now atm if you enable auto-game-time-ratio in server then in client the choosing lap will be setvisible false, and you vote the server will tell you the selected auto lap, so nothing you need to worry about

**Comment by hiker** [comment]
I like that, it solves all translations problems 👍

**Comment by Benau** [comment]
Also please merge master because your branch is 18x commits behind

**Comment by hiker** [comment]
> Also please merge master because your branch is 18x commits behind
Done, but mostly untested.

**Comment by hiker** [comment]
OK, I've removed me from this ticket, @benau, @deveee (or whoever), please take the ticket and add the voting list as you suggested.

**Comment by Benau** [comment]
maybe deve just let me handle the gui so u can fix the skidding or just git revert my commit

**Comment by Benau** [comment]
the removal of majority agree end early breaks my voting theory, because my screen is discussion-agreement-oriented, which picks the best race settings for players based on their discussion.

your random picking is good for your old approach, where agreement is not imortant, because it's copied from mario kart.

if you don't want different tracks / laps / reverse from different votes when they don't agree, just fix it, it's easy.

btw deve currently it "offsets" more if u choose cocoa / grand island! (seems that if screenshot size differ it has different behaviour)

and hiker "Could you let me know what target date you would prefer for a release" you didn't mention "release == beta / final!"

**Comment by Alayan-stk-2** [comment]
A few things : 
- In practice, you can't really discuss the choice. There is not enough time between the race, and there is not enough time nor chat to discuss it during the track selection itself.
- The first proposal by hiker meant that when doing your choice, you couldn't see what others had picked. Also, editing your choice took too much effort. But with deveee's suggestion, you get that feedback quickly. You want to match the choice of someone else ? Easy ! You want to put your own preference first ? Easy too ! The choice is left to the player.
- The removal of the "end early when majority agree" likely won't have much effect with many players (with 8 players, it's unlikely to get quickly a 5-players agreement on a setup), and for low-player count (3-4 players) you can still hope to have everyone agreeing on one choice, but without 2 players always rejecting what a 3rd want.
- Mixing and matching lap count/reverse/track from several votes is a bad idea. If someone wants math class and someone wants black forest, having 2 laps of math class or 6 laps of black forest is bad. In the same way, if one player wants snow peak and the other nessie's pond in reverse, giving them snow peak in reverse is bad.

**Comment by deveee** [comment]
It should be better now. Please check if I didn't break any other list, but as far as I see it's ok and kart icons in ghost replay list are now in correct position too.

I'm not sure what about word wrap in voting list. It should help for players with very long names, but even 3-letters "Yes" in reverse may use word wrap on lower resolutions and other translations may be longer. Maybe use reverse icon instead of "Yes" text, or maybe just "Y" or "N" ?

**Comment by hiker** [comment]
> if you don't want different tracks / laps / reverse from different votes when they don't agree, just fix it, it's easy.

No, I have already fixed that. Work on the voting gui should be based on the branch I have created and cleaned up. It contains all the logic required, only the list display is missing and needs to be added back. Feel free to cherry-pick code if this is more convenient. I won't accept any code that requires the old, tuple-based voting handling.

[the last item you mention seems not to belong to this ticket]

**Comment by Alayan-stk-2** [comment]
I tested the branch. Nicely done with the centered icon in list ! It also work now in the ghost replay selection list. :slightly_smiling_face: 

I don't think that word wrap is needed/good in the voting list, but maybe getting slightly more width would be useful. However, as far as I know, we can't make the proportion change depending on resolution ?

One small issue I noticed : the track pictures are not square (4:3 iirc), hence they appear deformed in the current voting UI list. Displayed at the correct ratio, they would look better.

**Comment by Benau** [comment]
I cannot understand why you hate majority agreement so much, if that's the case in the same logic you should againist the server owner changing game mode or start the race too, because that's 1 people force all

and with my way players can avoid some stupid choice from a player like 10 laps of cocoa (both long and slow track)

**Comment by deveee** [comment]
I don't care much about end voting early, because it's just that you save several seconds and it never happened for me in real voting anyway. But still I see nothing wrong here. You don't want it probably because it has a sense only if you use votes from majority?

And avoid some stupid choice is very good point. If I would be forced by one guy to play 20 laps, I would just leave such race.

**Comment by Alayan-stk-2** [comment]
In practice, I find the majority agreement triggers quite often. I played 25 games in the beta yesterday, with 3 to 5 players, and almost all had the majority agreement triggering, because very often we were aligning our vote with each other (we did not want too much to do again a previously done track, usually, so the vote was mostly about the order in which we did tracks).

So we could keep a "all agree ends early".

The random pick like was agreed upon last month has some upside (right now, it can be quite hard to get a track played in reverse if you're the only one wanting to try it, and it work better with many players than the current non-random pick) but the "one guys force other to do 20 laps" downside needs to be addressed. We had already discussed potential solutions, like server setting limiting to up twice the default lap number of the track, and one of them has to be implemented.

**Comment by Benau** [comment]
I don't understand, if you want to go extreme and let's go even further:

First, there could be some misunderstanding, you start by saying improving the voting gui, with only mentioning gui, using track/lap/reverse from one vote only, floating-point time bar... for the "random" part the closest thing i ever seen is the randomly picking track animation as seen in irc (which disabled atm), and imo whichever way (random / majority picking) can fit that animation

Next, the remove majority commit is "hidden" in your branch until i realized it and email you, and then you ask me to ask here for feedback here, and the result seems that no one cares or having a strong opinion. Finally (most important), you said "Feel free to cherry-pick code if this is more convenient", and you seem to forget / didn't see my message in december irc about
```
01:56 < Benau> hiker i really need the "end early when majority agree" 
01:56 < Benau> remove it is a regression
```
and your ticket is about improving the gui, so i think we should just change the gui but not anything else.
I may have taken your "this is more convenient" wrong, where this == old voting picking in my mind, if that's the case i feel regreted and apologize.

I don't want to comment on your criticism on majority voting agreement, if you are nip-picking you always find something. This is the same for pull request review, how you feel when you recieved alayan feeback for the powerup manager review, this time he nippicking you, next time the opposite, and then you wasting time criziting each other and no development will actually happen.

This is what i meant trust, for example there is not trust between humanity if you count the change returned by shop everytime.

And if you was just upset about I voted your 5 laps reversed stk enterprise out with alayan (seen in blogspot), I can tell you for next game if you still play, I (or others) will vote your choice next time, because that's how human with emotion works.

In short, you have wasted (long) time writing a long essay criticising me for what I could have been wrong, but right now there is more important thing for you to deal with, and I cannot see how it can improve the current situation.

**Comment by qwertychouskie** [comment]
Another problem:  If all the votes are different, the first vote always wins.

**Comment by hiker** [comment]
Since no one voted as I had requested from the core team, and considering future development, I'll just leave this issue for the future and not as a blocker for the next release.

**Comment by Arthur-D** [comment]
One issue I have with majority voting is that with majority voting, there's a risk of two or three people always rotating between a very limited number of tracks, because those are the tracks they are good at, and since they are the majority they get their track chosen every time. Also it seems like people want others to conform, telling them to vote a certain way, because then the voting period ends early with their preferred choice. Randomly selecting a set of settings avoids both of those, so I am in favor of doing that instead.

**Comment by hiker** [comment]
One vote ... hooray :)

Let me throw a curve ball: what if we would not allow to select the number of laps? But instead only allow the default number?  Do we see a lot of significant higher (or lower) number of laps selected in actual play?

[I am not saying that is the way to go, just try to get some brainstorming done ... though this would solve the one mentioned disadvantage that a single player selecting 20 laps can ruin a game if a single vote is selected]
